src/dbutils/mutation.rs

```diff
@@ -0,0 +1,359 @@
+use crate::{
+    kv::{
+        traits::{Cursor, MutableCursor},
+        CustomTable, Table,
+    },
+    MutableTransaction,
+};
+use akula_table_defs::TABLES;
+use anyhow::Result;
+use bytes::Bytes;
+use std::{
+    collections::{BTreeMap, BTreeSet, HashMap},
+    marker::PhantomData,
+};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+enum MutationError {
+    #[error("Table not found.")]
+    TableNotFound,
+}
+
+type TableName = string::String<static_bytes::Bytes>;
+type SimpleBucket = BTreeMap<Vec<u8>, Option<Vec<u8>>>;
+type SimpleBuffer = HashMap<TableName, SimpleBucket>;
+type DupSortValues = BTreeSet<Vec<u8>>;
+
+#[derive(Default)]
+struct DupSortChanges {
+    insert: DupSortValues,
+    delete: DupSortValues,
+}
+
+type DupSortBucket = BTreeMap<Vec<u8>, DupSortChanges>;
+type DupSortBuffer = HashMap<TableName, DupSortBucket>;
+
+fn is_dupsort(table_name: &str) -> bool {
+    if let Some(info) = TABLES.get(table_name) {
+        info.dup_sort.is_some()
+    } else {
+        false
+    }
+}
+
+struct Mutation<'tx: 'm, 'm, Tx: MutableTransaction<'tx>> {
+    parent: &'m mut Tx,
+    phantom: PhantomData<&'tx i32>,
+    simple_buffer: HashMap<TableName, SimpleBucket>,
+    dupsort_buffer: HashMap<TableName, DupSortBucket>,
+    sequence_increment: HashMap<TableName, u64>,
+}
+
+impl<'tx: 'm, 'm, Tx: MutableTransaction<'tx>> Mutation<'tx, 'm, Tx> {
+    fn new(parent: &'m mut Tx) -> Self {
+        Mutation {
+            parent,
+            phantom: PhantomData,
+            simple_buffer: Default::default(),
+            dupsort_buffer: Default::default(),
+            sequence_increment: Default::default(),
+        }
+    }
+
+    async fn get<T>(&'m self, table: &T, key: &[u8]) -> Result<Option<Bytes<'m>>>
+    where
+        T: Table,
+    {
+        let table_name = table.db_name();
+
+        if let Some(bucket) = self.simple_buffer.get(&table_name) {
+            Ok(match bucket.get(key) {
+                Some(entry) => entry.as_ref().map(|value| value.as_slice().into()),
+                None => self.parent.get(table, key).await?,
+            })
+        } else if let Some(_bucket) = self.dupsort_buffer.get(&table_name) {
+            todo!()
```

Comment:
I'm not sure what `get()` is supposed to return here - the first value, a random one, an `Err`? Also since there is no cursor implementation, I suppose we will also need a `get_dupsort()` returning all values for a key?

Reply:
I think it makes sense to replicate MDBX behaviour here, which would be get the first value.